Objects/listobject.c

```diff
@@ -2306,7 +2306,12 @@ list_sort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
 
             if (key->ob_type != key_type) {
                 keys_are_all_same_type = 0;
-                break;
+                /* If keys are in tuple we must loop over the whole list to make
+                   sure all items are tuples */
+                key_type = key->ob_type;
```

Comment:
It is not obvious which this assignment is needed. It is needed for the following checks for key_type, but that checks are not needed if keys_are_all_same_type == 0.

I would either add `continue` here, or add `else` before `if (key_type == &PyLong_Type)`.

Reply:
Hi @serhiy-storchaka, I added an replaced this by an explicit check on `keys_are_all_same_type` to make the logic less confusing.